### PR TITLE
[wallet] [rpc] getreceivedbyaddress should return error if called with address not owned by the wallet

### LINKED ISSUE #10921: getreceivedbyaddress should return error for non-IsMine addresses
Currently, the `getreceivedbyaddress` RPC will return a value of 0 if a non-IsMine address is passed in. This is confusing, and should probably return an error, as it's a situation the caller should be aware of.

Not doing this leads to confusion, with people assuming it can query non-wallet addresses.

See also https://bitcoin.stackexchange.com/questions/57065/why-bitcoin-cli-getreceivedbyaddress-always-returns-0

### PR DESCRIPTION
**Pull request by jnewbery**
Two commits:

- First commit tidies up the `receivedby.py` test (and speeds it up by factor of two)
- Second commit changes getreceivedbyaddress to return error if the address is not found in wallet, and adds test to `receivedby.py`

### CODE REVIEW

**Comment by jnewbery** [comment]
Should be included in v0.16 release notes

**Comment by jonasschnelli** [comment]
utACK 672fbde065df795d2d23080f2eec651c25875cb2.

**Comment by laanwj** [comment]
Concept ACK, this makes sense, there is no reason to ever call this on an unknown (neither directly owned nor watch-only) address.

**Comment by MeshCollider** [comment]
utACK https://github.com/bitcoin/bitcoin/commit/672fbde065df795d2d23080f2eec651c25875cb2

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -642,7 +642,7 @@ UniValue getreceivedbyaddress(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid Bitcoin address");
     CScript scriptPubKey = GetScriptForDestination(address.Get());
     if (!IsMine(*pwallet, scriptPubKey)) {
-        return ValueFromAmount(0);
+        throw JSONRPCError(RPC_WALLET_ERROR, "Address not owned by this wallet");
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]:
```cpp
throw JSONRPCError(RPC_WALLET_ERROR, "Address not found in wallet");
```

**Review by promag — COMMENTED** [review]
IMO these 2 commits would do:
- `Cleanup receivedby.py`
- `Throw error in getreceivedbyaddress if address is not found in wallet`

Where the 2nd commit adds the new error and also the corresponding test.

Otherwise tested ACK 672fbde.

**Comment by jnewbery** [comment]
Thanks @promag - both good suggestions.

I've squashed this down to 2 commits and changed the error text to "Address not found in wallet". No other changes.

**Comment by promag** [comment]
reACK 828c454.

**Comment by luke-jr** [comment]
Needs rebase

**Comment by jnewbery** [comment]
Rebased

**Comment by MarcoFalke** [comment]
@jnewbery Needs rebase

**Comment by jnewbery** [comment]
@MarcoFalke - apologies. I didn't realise there was a silent merge conflict. Rebased.

**Comment by MarcoFalke** [comment]
```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

utACK 5e0ba8f8cd55a5c2cea4e4e2343d7dcd52aa8fad
-----BEGIN PGP SIGNATURE-----

iQIcBAEBCgAGBQJaBzUWAAoJENLqSFDnUoslDMQQALgoBpiT9S6/rTIaOLTu1rHH
EgZIta8nFZtKL7jBE5t2hac2kC7DolgslTHcxFrblugEM2qth9eArUYbTZ+nSEQU
/8/N6rj3sQ/n1yumT3XGXLkdEI7JJNkq+mpeaw+e7iE4LwJ8rf3lZrVlbXp2vbtY
W333L8lJn8z3D5sbNJq6yXo/UgZIjM0bnkSoF4fmBhBhhI2uD4G/Vc1nIUIiSsF/
x6dUIcpAOp0/qfkqEVDKseKB2+zX3I+S1GVXxrafekvDPlcPNV5n4EXVJt6TSJP5
d5bfamzS2eDYD5YcxrvcXrty+z/T8cU/Je3ZSL5RoAiHgF/xJ+o0IjyCnERDXx5f
g5ax5nUuOhFfHd9xDukhjjyEUbgEqU7poYfT1LKeKU6jFIzE1BQl7xfbnV1tw57R
Yakj3rHgs15SGou26BxIuSazw+qT59IsmFDAzzOY0ur/iaYCvBmbsCU3UBq8YZ14
mf35nLlscDmgZhv8gMTF9OYK6y7Z7o9/JlEDZfxSo1/EYI+TG1LgiYcsTnzWN1js
L7bsZ704Pr1/XxSoILbM3t9wLeGXSRMlhfq2SWnWEfHHKFbtup06azs5pP5WCwEs
0JBc6Z6GBFCkio+fPdB/nc4LWULj6nWxRi2EFRZtWsMiJIo386RS/8+j0sKAYeHM
AG2KYNZuV4dABLTwnpO8
=CY9W
-----END PGP SIGNATURE-----

**Pull request merged**